common/sgx/verifier.c

```diff
@@ -302,98 +339,78 @@ static oe_result_t _fill_with_known_claims(
-            entry->name,
-            entry->name_size,
-            entry->name + entry->name_size,
-            entry->value_size));
-
-        // Go to next entry.
-        claims_buf += size;
-        claims_buf_size -= size;
-    }
+    // Default hash for empty string, as described in the literature.
+    static uint8_t sha256_for_empty_string[] = {
+        0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14, 0x9a, 0xfb, 0xf4,
+        0xc8, 0x99, 0x6f, 0xb9, 0x24, 0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b,
+        0x93, 0x4c, 0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55};
+
+    // Produce a hash of the custom_claims.
+    // If there is no data, set to default value
+    if (!custom_claims || !custom_claims_size)
+        memcpy(hash_out, sha256_for_empty_string, sizeof(*hash_out));
+    else
+        // Produce a hash of the custom_claims.
+        OE_CHECK(oe_sha256(custom_claims, custom_claims_size, hash_out));
 
     result = OE_OK;
 
 done:
-    if (result != OE_OK)
-    {
-        for (size_t i = 0; i < claims_index; i++)
-            _free_claim(&claims[i]);
-    }
     return result;
 }
 
 oe_result_t oe_sgx_extract_claims(
+    const sgx_evidence_format_type_t format_type,
     const oe_uuid_t* format_id,
-    const uint8_t* evidence,
-    size_t evidence_size,
+    const uint8_t* report_body,
+    size_t report_body_size,
+    const uint8_t* custom_claims,
+    size_t custom_claims_size,
     const oe_sgx_endorsements_t* sgx_endorsements,
     oe_claim_t** claims_out,
     size_t* claims_length_out)
 {
     oe_result_t result = OE_UNEXPECTED;
-    oe_report_header_t* header = (oe_report_header_t*)evidence;
-    oe_sgx_plugin_claims_header_t* claims_header = NULL;
-    size_t report_size = sizeof(*header) + header->report_size;
     oe_claim_t* claims = NULL;
     uint64_t claims_length = 0;
     uint64_t claims_size = 0;
     size_t claims_added = 0;
+    size_t additional_claim = 0;
+    sgx_report_data_t report_data;
 
-    // Check if the buffer is the proper size.
-    if (evidence_size - report_size < sizeof(*claims_header))
+    if (!format_id || !report_body || !report_body_size)
         OE_RAISE(OE_INVALID_PARAMETER);
 
-    // verify the integrity of the serialized claims with hash stored in
-    // report_data.
-    OE_CHECK(_verify_claims_hash(
-        header->report,
-        header->report_type,
-        evidence + report_size,
-        evidence_size - report_size));
-
-    claims_header = (oe_sgx_plugin_claims_header_t*)(evidence + report_size);
+    // verify the integrity of the custom_claims with hash stored in
+    // report_body, or retrieve the SGX report data
+    OE_CHECK(_process_sgx_report_data(
+        format_type,
+        report_body,
+        custom_claims,
+        custom_claims_size,
+        &report_data));
+
+    // There is no "custom_claims" claim if custom_claims
+    // is expected but is empty. Otherwise there is either a "custom_claims"
+    // claim or an SGX report data claim
```

Comment:
Consider making the text less confusing:
```suggestion
    // If provided, custom_claims can only be the OE_CLAIM_CUSTOM_CLAIMS
    // or OE_CLAIM_SGX_REPORT_DATA types. 
```

Reply:
Thank you!